wgengine/magicsock: fix BenchmarkReceiveFrom.

danderson opened the pull request:
Previously, this benchmark relied on behavior of the legacy
receive codepath, which I changed in 22507adf. With this
change, the benchmark instead relies on the new active discovery
path.

Signed-off-by: David Anderson <danderson@tailscale.com>

bradfitz reviewed (approved):
How about adding a hook (some package global func or bool) so the legacy path will panic if used? Then we don't have to worry about this silently breaking in the future if things get refactored.

This exact set of criteria of when to use disco is very convoluted to have mirrored in different places (I think the 2 places we had before was already bad, and 3 is worse).

So if we're going to mirror it again, at least we can add a safety mechanism?

danderson commented:
Added a conn option to disable legacy networking, which errors on legacy endpoint creation and refuses to do legacy anything, and updated the benchmark func to set it.

Can't set it across the board in tests yet, because it turns out some of our tests rely on legacy behavior :( Will fix those separately now that I have the knob for it.
